# FIXED COUPON BOND - QL EXAMPLE

This is a based on example in http://gouthamanbalaraman.com/blog/quantlib-bond-modeling.html

In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [3]:
from financepy.utils import *
from financepy.products.bonds.bond import *
from financepy.market.curves.discount_curve_zeros import *

# Define the Bond

In [4]:
issue_date = Date(15, 1, 2010)
maturity_date = Date(15, 1, 2016)
coupon = 0.06
freq_type = FrequencyTypes.SEMI_ANNUAL
accrual_type = DayCountTypes.THIRTY_360_BOND
face = ONE_MILLION

In [5]:
bond = Bond(issue_date, maturity_date, coupon, freq_type, accrual_type, face)

In [6]:
print(bond)

OBJECT TYPE: Bond
ISSUE DATE: 15-JAN-2010
MATURITY DATE: 15-JAN-2016
COUPON: 0.06
FREQUENCY: FrequencyTypes.SEMI_ANNUAL
ACCRUAL TYPE: DayCountTypes.THIRTY_360_BOND
FACE AMOUNT: 1000000


To see the cash flows we first need to set the settlement date of the bond. 

In [7]:
settlement_date = Date(15, 1, 2015)

In [8]:
bond.print_flows(settlement_date)

 15-JAN-2015      30000.00 
 15-JUL-2015      30000.00 
 15-JAN-2016    1030000.00 


## Discounting Bond Flows

We wish to define a zero rate curve. For this we need the dates and values of the zero rates.

In [9]:
zero_dates = [Date(15,1,2015), Date(15,7,2015), Date(15,1,2016)]
zero_rates = [0.00, 0.005, 0.007]

In [10]:
discount_curve = DiscountCurveZeros(settlement_date, zero_dates, zero_rates, 
                                      FrequencyTypes.ANNUAL,
                                      DayCountTypes.THIRTY_360_BOND)

In [11]:
print(discount_curve)

OBJECT TYPE: DiscountCurveZeros
VALUATION DATE: 15-JAN-2015
FREQUENCY TYPE: FrequencyTypes.ANNUAL
DAY COUNT TYPE: DayCountTypes.THIRTY_360_BOND
INTERP TYPE: InterpTypes.FLAT_FWD_RATES
DATES: ZERO RATES
 15-JAN-2015:  0.0000000
 15-JUL-2015:  0.0050000
 15-JAN-2016:  0.0070000



In [12]:
discount_curve._times

array([0. , 0.5, 1. ])

In [13]:
discount_curve._dfs

array([1.        , 0.99750934, 0.99304866])

In [14]:
bond.print_flows(settlement_date)

 15-JAN-2015      30000.00 
 15-JUL-2015      30000.00 
 15-JAN-2016    1030000.00 


In [15]:
discount_curve.df(settlement_date)

1.0

In [16]:
bond._flow_amounts

[0.0, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03]

In [17]:
bond.calc_accrued_interest(settlement_date)

30000.0

In [18]:
bond.clean_price_from_discount_curve(settlement_date, discount_curve)

105.27653992490683

In [19]:
bond.full_price_from_discount_curve(settlement_date, discount_curve)

108.27653992490683

As we are on the issue date of the bond there is a full coupon of accrued.

This agrees with QL which finds a clean price of 105.27654

Copyright (c) 2020 Dominic O'Kane